aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -685,6 +686,434 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+      6.75961301770459654925e-02,
+      2.95552570429315521194e-02,
+      -1.39180257516060693404e-05
+    },
+    {
+      0.,
+      0.,
+      -9.23364619367118555360e-02,
+      -1.69364939002081722752e-02,
+      -1.40086798182036094347e-05
+    }
+  }};
+
+  // gather coefficients `b` from above into a tensor,
+  // and move them to device of type `A.device().type()`
+  auto bs = at::from_blob(
+    reinterpret_cast<void*>(&b),
+    {num_prods, num_prods},
+    {num_prods, 1},
+    A.dtype()
+  ).to(A.device().type());
+
+  auto As = _allocate_buffer(A, num_prods);
+  _fill_matrix_powers(As, A, num_prods);
+
+  auto Bs = at::native::_compute_linear_combination(As, bs);
+
+  const auto A9 = at::matmul(Bs.select(0, 0), Bs.select(0, 4)) + Bs.select(0, 3);
+  const auto res = Bs.select(0, 1) + at::matmul(Bs.select(0, 2) + A9, A9);
+  return res;
+}
+
+template <typename scalar_t>
+Tensor mexp_impl(
+  const Tensor& a,
+  std::array<scalar_t, total_n_degs> thetas,
+  bool compute_highest_degree_approx = false
+) {
+  auto res = at::empty_like(a);
+  const auto norm = operator_1_norm(a);
+
+  if (!compute_highest_degree_approx) {
+    constexpr std::array<
+      Tensor(*)(const Tensor&),
+      total_n_degs - 1> 
+    compute_Ts = {
+      compute_T1, compute_T2, compute_T4,
+      compute_T8<scalar_t>, compute_T12<scalar_t>
+    };
+
+    for (int i = 0; i < total_n_degs - 1; ++i) {
+      auto norm_lower_bound = (i == 0) ? static_cast<scalar_t>(-1) : thetas[i - 1];
+      auto norm_upper_bound = thetas[i];
+      // nonzero returns a 2D tensor, hence squeeze(-1) to make it 1D
+      auto idx_curr_norm_interval = (
+        (norm_lower_bound < norm) * (norm <= norm_upper_bound)
+      ).nonzero().squeeze(-1);
+
+      if (idx_curr_norm_interval.numel() > 0) {
+        auto sub_a = at::index_select(a, 0, idx_curr_norm_interval);
+        res.index_put_({idx_curr_norm_interval}, compute_Ts[i](sub_a));
+        return res;
+      }
+    }
+
+    auto idx_large_norm = (norm >= thetas[total_n_degs - 1])
+      .nonzero().squeeze(-1);
+    if (idx_large_norm.numel() > 0) {
+      auto a_large_norm = at::index_select(a, 0, idx_large_norm);
+      auto large_norm_subset = at::index_select(norm, 0, idx_large_norm);
+
+      // Scale
+      const auto s = at::max(
+        at::zeros_like(large_norm_subset),
+        at::ceil(at::log2(large_norm_subset / thetas[total_n_degs - 1]))
+      ).unsqueeze(-1).unsqueeze(-1).to(at::kLong);
+      const auto pow2s = at::pow(2, s);
+      const auto a_large_norm_scaled = a_large_norm / pow2s;
+
+      // Square
+      auto mexp_scaled = at::native::compute_T18<scalar_t>(a_large_norm_scaled);
+      auto mexp_buffer = at::empty_like(mexp_scaled);
+      for (int64_t i = 0; i < mexp_scaled.size(0); ++i) {
+        auto mexp = mexp_scaled.select(0, i);
+        while ((s.select(0, i) > 0).nonzero().numel()) {
+          mexp = at::matmul(mexp, mexp);
+          s.select(0, i) -= 1;
+        }
```

Comment:
@ngimel, could you please tell whether this part actually synchronize?

Reply:
What is the best way to check whether tensor is empty?

Reply:
`nonzero` synchronizes. Any data dependent behavior that causes host side code to change behavior, by necessity, causes a synchronization; if you want to eliminate the synchronization, you need to figure out how to phrase your algorithm in a data oblivious way.

Reply:
I see, thank you! And when I move data from CPU to CUDA with `.to(at::kCUDA)`, does it always happen asynchronously?

Reply:
when you move data from cpu to gpu with `.to` it will synchronize, unless your cpu tensor is pinned.

Reply:
@ngimel, is it unconditional on whether `non_blocking` is set?

Reply:
`non-blocking` has a real effect only if cpu memory is pinned. If it's not, then what happens under the hood is first the data is synchronously copied to the pinned buffer, and then, if `non_blocking` is set, asynchronously copied from pinned buffer to device, so the end effect is still almost as if you had full synchronization, see https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html